pre-commit/check.py

```diff
@@ -0,0 +1,161 @@
+    git_status = os.popen("git status --porcelain").read()
+    git_status_list = git_status.split("\n")
+    git_status_list = [x for x in git_status_list if x]
+    files = [Path(x.split()[1]).absolute() for x in git_status_list]
+    return files
+
+
+# gets last changed files
+def get_changed_files() -> List[Path]:
+    files = get_git_status()
+    if len(files) != 0:
+        return files
+    files = get_git_commit_files("HEAD")
+    return files
+
+
+# check if path a child of another path
+def is_child(path: Path, parent: Path) -> bool:
+    assert path.is_absolute()
+    assert parent.is_absolute()
+    cur_path = path
+    while cur_path != parent and cur_path != root:
+        cur_path = cur_path.parent
+        if cur_path == parent:
+            return True
+    return False
+
+
+def unpushed_files() -> List[Path]:
+    """return all files that are committed but not pushed."""
+    # get upstream branch name
+    upstream_branch = (
+        os.popen("git rev-parse --abbrev-ref --symbolic-full-name @{u}").read().strip()
+    )
+    # get all commits that are on the current branch but not on the upstream branch
+    commits = os.popen(f"git log {upstream_branch}..HEAD --oneline").read().strip()
+    commits = [x.split()[0] for x in commits.split("\n")]
+    # get all files that are changed in the commits
+    files: List[Path] = []
+    for commit in commits:
+        files.extend(get_git_commit_files(commit))
+    # remove duplicates
+    return sorted(list(set(files)))
+
+
+# check if PRECOMMIT_ENABLE_SLOW is set to true
+if os.environ.get("PRE_COMMIT_ENABLE_SLOW", "false") != "true":
+    print("Skipping slow checks.")
+    print("Set PRE_COMMIT_ENABLE_SLOW=true to activate.")
+    exit(0)
+
+PROJECT_NAME = "determined"
+root = Path(os.getenv("PROJECT_ROOT", os.getcwd())).absolute()
+if not str(root).endswith(PROJECT_NAME) and not str(root).endswith("saas"):  # FIXME
+    print(
+        "Please run this script from the root of the project or set it using "
+        + "PROJECT_ROOT env variable"
+    )
+    exit(1)
+os.chdir(root)
+
+rules: Dict[Path, Union[str, List[str]]] = {
+    root
+    / "harness": [
+        "make -j fmt; make -j check",
+        "make -j build",
+    ],
+    root
+    / "proto": [
+        "make fmt check build",
+        "make -C ../bindings build && make -C ../webui/react bindings-copy-over && make -C ../webui/react check",
+    ],
+    root / "webui" / "react": ["make -j fmt; make -j check", "make -j test && make -j build"],
+    root
+    / "master": [
+        "make -C ../proto build",
+        "make build",
+        "make fmt; make check",
+        "make test",
+    ],
+    root / "docs": "make fmt check build",
+    root / ".circleci": "circleci config validate config.yml",
+    root / "e2e_tests": "make fmt check",
+    root / "model_hub": "make fmt check",
+}
+
+
+# check if path is the same or child of one of the rules and execute the rule as
+# subprocess
+def run_rule(rule_path: Path) -> int:
```

Comment:
How do you feel about making rule-checking run in-parallel?
Like, for each known rule we need to run, spinning a subprocess off and just capturing the rc, stdout, and stderr to display if rc != 0. Basically, fan-out to run all rules needed, then fan-in and print out a report of failures.

I don't normally care for these kinds of optimizations without good reason, but this _is_ gonna be running pretty frequently on a local machine, so it'd be prudent to make it as unobtrusive as possible.

Reply:
do we want to bail out as quickly as possible or lean toward running more checks?

Reply:
say you have a 10sec check and a 1minute check but your issue is with the 10sec check. with pre-commit I donn't know if we can provider early feedback without exiting the (check.py) process with a non-zero code

Reply:
I updated the runner with a simple multiprocessing strategy. I think I want to start using it to see where the friction points are and what behavior would work best